src/meshcat/jupyter.py

```diff
@@ -0,0 +1,80 @@
+"""
+A Visualizer for use in Jupyter notebooks or Google's Colaboratory.  It uses ipykernel.comm instead of zmq + websockets to establish a 1:1 connection
+between python and the javascript viewer.  This was necessary, because colab
+blocks insecure websockets (ws), and it is hard to establish and accept a
+self-signed certificate for every cloud instance.
+
+See https://github.com/RobotLocomotion/drake/issues/12645 for the detailed
+origin story.
+"""
+
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+import os
+import sys
+from IPython.display import HTML, display
+from ipykernel import comm
+
+from .path import Path
+from .commands import SetObject, SetTransform, Delete, SetProperty, SetAnimation
+
+
+class JupyterVisualizer:
+    __slots__ = ["path"]
+
+    def __init__(self, write_html=True):
+        self.path = Path(("meshcat",))
+        if write_html:
+            main_min = os.path.dirname(__file__) + '/viewer/dist/main.min.js'
+            with open(main_min, "r") as f:
+                main_min_js = f.read()
+
+            comm_js = 'window.parent.Jupyter.notebook.kernel.comm_manager'
+            if 'google.colab' in sys.modules:
+                comm_js = 'google.colab.kernel.comms'
+            display(HTML(f"""
+<div id="meshcat-pane" style="height: 400px; width: 100%; overflow-x: auto; overflow-y: hidden; resize: both">
+</div>
+
+<script type="text/javascript">
+{main_min_js}
```

Comment:
Sorry for the pushes.  I've generalized it so that it works on colab and in jupyter notebooks more generally, except that this trick seems to not quite be working for me locally.  I'm getting 
```
Uncaught ReferenceError: MeshCat is not defined
```
Any ideas?

Reply:
Yes. You just have to put the whole html code in a iframe manually. Do not forget to properly escape a few special characters:
```
            main_min = os.path.dirname(__file__) + '/viewer/dist/main.min.js'
            with open(main_min, "r") as f:
                main_min_js = f.read()
                main_min_js = main_min_js.replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;").replace("'", "&apos;")

            comm_js = 'window.parent.Jupyter.notebook.kernel.comm_manager'
            if 'google.colab' in sys.modules:
                comm_js = 'google.colab.kernel.comms'
            display(HTML(f"""
<iframe srcdoc='          
<div id="meshcat-pane" style="height: 400px; width: 100%; overflow-x: auto; overflow-y: hidden; resize: both">
</div>
<script type="text/javascript">
{main_min_js}
</script>
<script>
var viewer = new MeshCat.Viewer(document.getElementById("meshcat-pane"));
{comm_js}.registerTarget("meshcat_comms", (comm, message) => {{
    viewer.handle_command(message.data)
}});
</script>'; style="width: 100%; height: 100%; border: none">
"""))
```

Reply:
Yet I get another error: `Uncaught TypeError: window.parent.Jupyter.notebook.kernel.comm_manager.registerTarget is not a function`

Reply:
Ok it is because it should be `register_target` instead of `registerTarget` for Jupyter. But it is still not working x)

Reply:
thanks for all of the help.  you got me unstuck.  i've got it all working now.

Reply:
message.data => message.content.data 

and I still need to block the sender til the browser is listening.

Reply:
well, that turned into a total mess.

it works beautifully on colab.  it works on my local jupyter notebook, but is clunky.  i've added a note to the file docstring.  @duburcqa -- if you manage to figure it out, that would be awesome!!

@rdeits -- this is my current best offering.  I would like the name to be `jupyter` instead of `colab`, because i have to believe that it will work for both.  but I also really want to move forward with using this on colab, since that's the primary use case.  If you like, I could throw a warning on non-colab... but that would be annoying for users who are making it work despite the handicap.

PTAL?

Reply:
Ok so after spending a few more hours looking at it, it seems that `sleep` cannot be avoided because `display` is not waiting for the script to complete, and there is no way to enforce it. I would recommand to use something much smaller though, such that `0.2` (on my compute, 0.04 is ok). Also I did some polishing to improve the rendering on Jupyter:
```
            display(HTML(f"""
<div style="height: 400px; width: 100%; overflow-x: auto; overflow-y: hidden; resize: both; border: none">
<iframe scrolling="no"; style="height: 100%; width: 100%; border: none"; srcdoc='        
<html>
<body><div id="meshcat-pane" style="height: 100%; width: 100%; border: none"></div></body>
<script>
window.parent.Jupyter.notebook.kernel.execute("meshcat_ready = False")
</script>
<script type="text/javascript">
{main_min_js}
</script>
<script>
var viewer = new MeshCat.Viewer(document.getElementById("meshcat-pane"));
window.parent.Jupyter.notebook.kernel.comm_manager.register_target("meshcat_comms", function(comm, msg) {{
    viewer.handle_command(msg.content.data);
}});
window.parent.Jupyter.notebook.kernel.execute("meshcat_ready = True")
</script>
</html>
</div>'>
"""))
```

Reply:
Thanks!  Were you able to use the "meshcat_ready = True" to block?  e.g. put a small sleep in a loop that terminates when you get meshcat_ready True?

Reply:
I tried using some global flag but it didn't work. The flag is never set if I do use a naive while/sleep loop to wait for it. It is only executed after the cell has returned apparently...